crates/api_server/src/error.rs

```diff
@@ -1,167 +1,87 @@
-use jsonrpsee::types::{ErrorCode, ErrorObjectOwned};
+use jsonrpsee::{
+    core::Serialize,
+    types::{ErrorCode, ErrorObject, ErrorObjectOwned},
+};
 use zksync_error::{
-    anvil_zksync::{node::AnvilNodeError, rpc::RpcError, state::StateLoaderError},
-    core::web3::Web3Error,
-    CustomErrorMessage as _, ICustomError as _, IError as _,
+    anvil_zksync::{node::AnvilNodeError, state::StateLoaderError},
+    ICustomError, IError as _, ZksyncError,
 };
-use zksync_types::{L1BatchNumber, L2BlockNumber};
-use zksync_web3_decl::error::Web3Error as LegacyWeb3Error;
+use zksync_web3_decl::error::Web3Error;
+
+use jsonrpsee::types::ErrorCode as RpcErrorCode;
 
-// TODO This datatype may be erased once we port Web3Error from ZKsync to
-// zksync-error
-#[derive(Debug, thiserror::Error)]
-pub enum JsonRpcAdapter {
-    LegacyWeb3(#[from] LegacyWeb3Error),
-    Node(#[from] AnvilNodeError),
-    Web3(#[from] Web3Error),
-    LoadState(#[from] StateLoaderError),
-    Anyhow(#[from] anyhow::Error),
+fn to_rpc<I: Serialize + ICustomError<ZksyncError, ZksyncError>>(
+    code: Option<i32>,
+    error: I,
+) -> ErrorObjectOwned {
+    ErrorObject::owned(
+        code.unwrap_or(error.to_unified().get_identifier().encode() as i32),
+        error.to_unified().get_message(),
+        Some(error),
+    )
 }
 
-impl std::fmt::Display for JsonRpcAdapter {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.write_fmt(format_args!("{self:#?}"))
-    }
+
+pub trait RpcErrorAdapter {
+    fn into(error: Self) -> ErrorObjectOwned;
 }
 
-impl From<JsonRpcAdapter> for ErrorObjectOwned {
-    fn from(val: JsonRpcAdapter) -> Self {
-        let rpc_error = match val {
-            JsonRpcAdapter::LegacyWeb3(web3_error) => RpcError::Web3Error {
-                inner: Box::new(legacy_web3_adapter(web3_error)),
-            },
-            JsonRpcAdapter::Node(anvil_node) => RpcError::NodeError {
-                inner: Box::new(anvil_node),
-            },
-            JsonRpcAdapter::Web3(web3) => RpcError::Web3Error {
-                inner: Box::new(web3),
-            },
-            JsonRpcAdapter::LoadState(state_loader) => RpcError::LoadStateError {
-                inner: Box::new(state_loader),
-            },
-            JsonRpcAdapter::Anyhow(error) => RpcError::GenericError {
-                message: error.to_string(),
-            },
-        };
-        into_json_rpc(rpc_error)
+impl RpcErrorAdapter for StateLoaderError {
+    fn into(error: Self) -> ErrorObjectOwned {
+        to_rpc(None, error)
```

Comment:
Doesn't this mean we use `StateLoaderError`'s internal code for JSON RPC responses?

Reply:
My bad, thanks for spotting.